components/validator/challenger.go

```diff
@@ -65,191 +93,413 @@ func NewChallenger(ctx context.Context, cfg Config, l log.Logger) (*Challenger,
+		l2ooABI:           l2ooABI,
 		colosseumContract: colosseumContract,
+		colosseumABI:      colosseumABI,
 
 		submissionInterval:        submissionInterval,
 		finalizationPeriodSeconds: finalizationPeriodSeconds,
 		l2BlockTime:               l2BlockTime,
-	}, nil
-}
 
-func (c *Challenger) IsChallengeInProgress() (bool, error) {
-	return c.colosseumContract.IsInProgress(c.callOpts)
+		txCandidatesChan: txCandidateChan,
+	}, nil
 }
 
-func (c *Challenger) GetChallengeInProgress() (bindings.TypesChallenge, error) {
-	return c.colosseumContract.GetChallengeInProgress(c.callOpts)
-}
+func (c *Challenger) Start() error {
+	c.log.Info("start challenger")
 
-func (c *Challenger) OutputAtBlockSafe(blockNumber uint64, includeNextBlock bool) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.NetworkTimeout)
-	defer cancel()
-	output, err := c.cfg.RollupClient.OutputAtBlock(ctx, blockNumber, includeNextBlock)
+	// set the challenger's checkpoint to the latest output index
+	// because when challenger starts, it is guaranteed that the past outputs are checked through scanPrevOutputs
+	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	c.checkpoint = new(big.Int).Sub(nextOutputIndex, common.Big1)
 
-	return output, nil
+	// scan all the outputs within the finalization window
+	c.wg.Add(1)
+	go c.scanPrevOutputs()
+
+	// if challenge mode on, subscribe L2 output submission events
+	if !c.cfg.ChallengerDisabled {
+		c.wg.Add(1)
+		go c.subscribeL2OutputSubmitted()
+	}
+
+	// subscribe challenge creation events
+	c.wg.Add(1)
+	go c.subscribeChallengeCreated()
+
+	return nil
 }
 
-type OutputRange struct {
-	OutputIndex *big.Int
-	StartBlock  uint64
-	EndBlock    uint64
+func (c *Challenger) Stop() {
+	c.log.Info("stop challenger")
+	close(c.done)
+	c.wg.Wait()
 }
 
-func (c *Challenger) GetInvalidOutputRange() (*OutputRange, error) {
-	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
-	if err != nil {
-		return nil, err
-	}
-	latestOutputIndex := new(big.Int).Sub(nextOutputIndex, common.Big1)
-	if latestOutputIndex.Cmp(common.Big1) == -1 {
-		c.log.Info("challenge for genesis output is not allowed", "latestOutputIndex", latestOutputIndex)
-		return nil, nil
+// scanPrevOutputs scans all the previous outputs within the finalization window.
+// If there are invalid outputs, create challenge.
+// If there are challenges in progress, keep handling them.
+func (c *Challenger) scanPrevOutputs() {
+	defer c.wg.Done()
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), c.cfg.NetworkTimeout)
+	defer cancelCtx()
+	status, err := c.cfg.RollupClient.SyncStatus(ctx)
+
+	toBlock := new(big.Int).SetUint64(status.CurrentL1.Number)
+	// fromBlock = max(1, startBlockOfFinalizationWindow)
+	fromBlock := math.BigMax(new(big.Int).Sub(toBlock, new(big.Int).Div(c.finalizationPeriodSeconds, big.NewInt(12))), big.NewInt(1))
```

Comment:
Please use `L1BlockTime` variable which is read from confing.

Reply:
The `L1BlockTime` is not in config.
When I add the `L1BlockTime` to rollup config, it makes many file changes.
Thus I mark `TODO` and will do in another PR.